From e1c5f24ba81fd093f47f705237806edc3a61011a Mon Sep 17 00:00:00 2001
From: Jaehyoung Choi <jkkkkk.choi@samsung.com>
Date: Tue, 18 Feb 2025 10:48:19 +0900
Subject: [PATCH] toolbox/modprobe: Add parallel option to load modules in parallel

Check the parallel option to determine
whether to call LoadListedModules or LoadModulesParallel.

The option applies when modules_load_file exists.

Bug: 404353675
Signed-off-by: Jaehyoung Choi <jkkkkk.choi@samsung.com>
Change-Id: I541e656f721f5e4c97acae442c597f4e4ae61a58
---

diff --git a/toolbox/modprobe.cpp b/toolbox/modprobe.cpp
index c5025e7..da252cb 100644
--- a/toolbox/modprobe.cpp
+++ b/toolbox/modprobe.cpp
@@ -20,6 +20,7 @@
 #include <unistd.h>
 
 #include <string>
+#include <thread>
 
 #include <android-base/file.h>
 #include <android-base/logging.h>
@@ -52,6 +53,7 @@
     LOG(INFO) << "  -D, --show-depends: Print dependencies for modules only, do not load";
     LOG(INFO) << "  -h, --help: Print this help";
     LOG(INFO) << "  -l, --list: List modules matching pattern";
+    LOG(INFO) << "  -p, --parallel: Load modules in parallel";
     LOG(INFO) << "  -r, --remove: Remove MODULE (multiple modules may be specified)";
     LOG(INFO) << "  -s, --syslog: print to syslog also";
     LOG(INFO) << "  -q, --quiet: disable messages";
@@ -157,7 +159,7 @@
     std::string mods;
     std::vector<std::string> mod_dirs;
     modprobe_mode mode = AddModulesMode;
-    bool blocklist = false;
+    bool blocklist = false, parallel = false;
     int rv = EXIT_SUCCESS;
 
     int opt, fd;
@@ -172,13 +174,14 @@
         { "show-depends",        no_argument,       0, 'D' },
         { "help",                no_argument,       0, 'h' },
         { "list",                no_argument,       0, 'l' },
+        { "parallel",            no_argument,       0, 'p' },
         { "quiet",               no_argument,       0, 'q' },
         { "remove",              no_argument,       0, 'r' },
         { "syslog",              no_argument,       0, 's' },
         { "verbose",             no_argument,       0, 'v' },
     };
     // clang-format on
-    while ((opt = getopt_long(argc, argv, "a::bd:Dhlqrsv", long_options, &option_index)) != -1) {
+    while ((opt = getopt_long(argc, argv, "a::bd:Dhlpqrsv", long_options, &option_index)) != -1) {
         switch (opt) {
             case 'a':
                 // toybox modprobe supported -a to load multiple modules, this
@@ -217,6 +220,9 @@
                 check_mode();
                 mode = ListModulesMode;
                 break;
+            case 'p':
+                parallel = true;
+                break;
             case 'q':
                 android::base::SetMinimumLogSeverity(android::base::WARNING);
                 break;
@@ -300,7 +306,10 @@
 
     Modprobe m(mod_dirs, modules_load_file.empty() ? "modules.load" : modules_load_file, blocklist);
     if (mode == AddModulesMode && !modules_load_file.empty()) {
-        if (!m.LoadListedModules(false)) {
+        bool retval = (parallel) ? m.LoadModulesParallel(std::thread::hardware_concurrency())
+                                 : m.LoadListedModules(false);
+
+        if (!retval) {
             PLOG(ERROR) << "Failed to load all the modules from " << modules_load_file;
             return EXIT_FAILURE;
         }
